File: GoTo/167.two-sum-ii-input-array-is-sorted.go
package leetcode

// TwoSum167 uses two pointers on the sorted array. A low sum moves left forward;
// a high sum moves right backward. Returned indices are 1-based.
//
// Time: O(n). Space: O(1).
func TwoSum167(numbers []int, target int) []int {
	l, r := 0, len(numbers)-1
	for l < r {
		sum := numbers[l] + numbers[r]
		if sum == target {
			return []int{l + 1, r + 1}
		}
		if sum < target {
			l++
		} else {
			r--
		}
	}
	return []int{}
}

File: companies/Instacart/167.two-sum-ii-input-array-is-sorted.py
"""
Approach: Two pointers from both ends of the sorted array.
Data structure: indexes only; sorted order replaces a hash table.
Interview logic: if the sum is too small, increasing the left pointer is the only way to make it larger. If too large, decreasing the right pointer is the only way to make it smaller.
Complexity: O(n) time, O(1) space.
Tests and edge cases: exactly one solution; negative numbers; answer indices are 1-based.
"""
from __future__ import annotations
from typing import List

# @lc code=start
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        left, right = 0, len(numbers) - 1
        while left < right:
            total = numbers[left] + numbers[right]
            if total == target:
                return [left + 1, right + 1]
            if total < target:
                left += 1
            else:
                right -= 1
        return []
# @lc code=end
